### DISCUSSION TITLE
Ability to move packages across repositories

### DISCUSSION THREAD

**Discussion by dentuzhik** [community/community#8577] [Actions] [answered]
Hey guys, thanks for a shiny new space for feedback 🎉 
It would be extremely useful to have an ability of moving a github packages across repositories, and have been brought up by several discussions in https://github.community:
* https://github.community/t/how-to-move-package-from-one-repository-to-another/137518
* https://github.community/t/move-npm-package-to-other-repo-package-is-already-associated-with-another-repository/2722/5
* https://github.community/t/can-i-move-a-nuget-package-to-a-new-repository/14832
* https://github.community/t/move-packages-from-one-repo-to-another/158312/2

Our cases are for this are specifically related to the fact that we are managing most of our code via monorepositories, and the following cases come out:
* Internal forks of open source libraries re-published in private registry, which we would like move to monorepositories and preserve the name (they can still be managed efficiently via `git subtree`)
* Merging packages into a single monorepo, merging several monorepos with packages together, or reverse splitting
* Restructuring teams and ownership of the packages, which sometime causes movement of the code across repositories

It is obvious that we can achieve this by renaming packages, however this quite a chore (and naming is hard) plus it causes non-zero educational overhead on the rest of the engineers which actually consume these packages, so if there would be a better way to change relationship to a parent repo, it would be nice.

Otherwise it would be interesting to hear rationale behind such strict package-repository lock, since other registries like NPM do not really have it.

**Comment by Hobart2967** [comment]
Hi @dentuzhik,

we had exactly the same problem over here. What we did:

Scripting a solution using Node.js + Axios + GitHub Rest Api:

- Pull information about packages (List of packages per repository)
- Download all packages and versions while getting the tarball URL using `npm view <package> dist.tarball`
- Untar files, adjust "repository.url" to the new repositories inside the package.json
- Delete all packages versions using the GitHub Rest API
- Upload the packages to the new repository using `npm publish``

That actually helped "moving" the packages :)

**Comment by maverick1872** [comment]
Why they're tightly associated to any repo other than for providing a easy link to it's source seems odd to me personally. Totally understand tightly coupling to the scope they're published under but I don't feel it should go further than that personally.

**Comment by eric-creekside** [comment]
More than 18 months later and no response. Is there a related feature request for this?

**Comment by shidil** [comment, marked as answer]
This is possible now by setting the [repository](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#repository) field in package.json

This is what worked for me

1. Changed package.json repository field to point to new repo `"repository": "git://github.com/org/new-repo.git"`
2. Published a new version of the package (edit package.json version and `npm publish`)
3. Went to package settings https://github.com/orgs/{orgName}/packages/npm/{packageName}/settings
4. Previous repo will be displayed, along with an unlink (delete icon) button. Unlink the package
     ![image](https://github.com/community/community/assets/4880359/824947e9-e5d9-4211-bb24-8c84a06a4228)
5. Published a new version (edit package.json version and `npm publish`)
6. Package moved to new repo!

_Maybe step 2 can be skipped_

**Reply by jschlechtehomer** [reply]
This worked for me. Thanks!

**Reply by fabiancook** [reply]
A missing step here if the package was previously only published through access, is to go to the package settings beforehand and add the target repository to the "Manage Actions access", then set the role for the new repository to Admin, the old repository can be set to write or read from there if needed too. 

<img width="341" height="229" alt="image" src="https://github.com/user-attachments/assets/b6210b27-c258-4b5b-8c81-22e467ce958e" />

The new repository can then do a publish through actions, and then the unlink button will appear and can be pressed to reset the repository source.

You may also want to change the package visibility from private to **internal** as well if you are going to have users read the package.
